### PR TITLE
Adding correctness in the ReplyCount generation

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Adding correctness in the ReplyCount generation. This way all the posts has the ReplyCount of the thread it belongs to.

This is part of the skip threads optimization.

#### Ticket Link
[MM-17468](https://mattermost.atlassian.net/browse/MM-17468)

### CODE REVIEW

File: store/sqlstore/post_store.go [hunk 1]
```diff
@@ -270,7 +275,7 @@ func (s *SqlPostStore) GetFlaggedPosts(userId string, offset int, limit int) (*m
 	pl := model.NewPostList()
 
 	var posts []*model.Post
-	if _, err := s.GetReplica().Select(&posts, "SELECT *, (SELECT count(Posts.Id) FROM Posts WHERE Posts.RootId = p.Id AND Posts.DeleteAt = 0) as ReplyCount FROM Posts p WHERE Id IN (SELECT Name FROM Preferences WHERE UserId = :UserId AND Category = :Category) AND DeleteAt = 0 ORDER BY CreateAt DESC LIMIT :Limit OFFSET :Offset", map[string]interface{}{"UserId": userId, "Category": model.PREFERENCE_CATEGORY_FLAGGED_POST, "Offset": offset, "Limit": limit}); err != nil {
+	if _, err := s.GetReplica().Select(&posts, "SELECT *, (SELECT count(Posts.Id) FROM Posts WHERE Posts.RootId = (CASE WHEN p.RootId = '' THEN p.Id ELSE p.RootId END) AND Posts.DeleteAt = 0) as ReplyCount FROM Posts p WHERE Id IN (SELECT Name FROM Preferences WHERE UserId = :UserId AND Category = :Category) AND DeleteAt = 0 ORDER BY CreateAt DESC LIMIT :Limit OFFSET :Offset", map[string]interface{}{"UserId": userId, "Category": model.PREFERENCE_CATEGORY_FLAGGED_POST, "Offset": offset, "Limit": limit}); err != nil {
```
Comment (reflog) [inline store/sqlstore/post_store.go]: `CASE WHEN p.RootId = '' THEN p.Id ELSE p.RootId END`. wow. everyday I learn something new :)

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -178,6 +178,11 @@ func (s *SqlPostStore) SaveMultiple(posts []*model.Post) ([]*model.Post, *model.
 			if ok {
 				post.ReplyCount += int64(count)
 			}
+		} else {
+			count, ok := rootIds[post.RootId]
+			if ok {
+				post.ReplyCount += int64(count)
+			}
 		}
```
Comment (reflog) [inline store/sqlstore/post_store.go]: Nit 0/5 I'd rewrite it as:
```suggestion
	for _, post := range posts {
	    id := post.RootId
		if len(post.RootId) == 0 {
     		id := post.Id
		}
		count, ok := rootIds[id]
		if ok {
			post.ReplyCount += int64(count)
		}
		
```

File: store/sqlstore/post_store.go [hunk 3]
```diff
@@ -825,10 +870,11 @@ func (s *SqlPostStore) getRootPosts(channelId string, offset int, limit int, ski
 	var posts []*model.Post
 	var fetchQuery string
 	if skipFetchThreads {
-		fetchQuery = "SELECT p.*, (SELECT COUNT(Posts.Id) FROM Posts WHERE p.RootId = '' AND Posts.RootId = p.Id AND Posts.DeleteAt = 0) as ReplyCount FROM Posts p WHERE ChannelId = :ChannelId AND DeleteAt = 0 ORDER BY CreateAt DESC LIMIT :Limit OFFSET :Offset"
+		fetchQuery = "SELECT p.*, (SELECT COUNT(Posts.Id) FROM Posts WHERE Posts.RootId = (CASE WHEN p.RootId = '' THEN p.Id ELSE p.RootId END) AND Posts.DeleteAt = 0) as ReplyCount FROM Posts p WHERE ChannelId = :ChannelId AND DeleteAt = 0 ORDER BY CreateAt DESC LIMIT :Limit OFFSET :Offset"
 	} else {
 		fetchQuery = "SELECT * FROM Posts WHERE ChannelId = :ChannelId AND DeleteAt = 0 ORDER BY CreateAt DESC LIMIT :Limit OFFSET :Offset"
 	}
+	mlog.Debug(fetchQuery, mlog.Any("params", map[string]interface{}{"ChannelId": channelId, "Offset": offset, "Limit": limit}))
```
Comment (hanzei) [inline store/sqlstore/post_store.go:877]: @jespino Did you leave this here by accident? Usually queries don't get logged.

Comment (jespino) [reply store/sqlstore/post_store.go:877]: Yes, I going to make a PR right away to remove that, sorry.

**Pull request merged**